libraries/stumbler/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/motiondetection/LegacyMotionSensor.java

```diff
@@ -48,57 +39,65 @@ public LegacyMotionSensor(Context ctx) {
         mSensorManager = (SensorManager) mAppContext.getSystemService(Context.SENSOR_SERVICE);
     }
 
-    /*
-     This function is called by the SensorEventListener so that we have something to test.
-     Real SensorEvent instances cannot be constructed on a test platform.
-     */
-    public void sensorChanged(int sensorType, float[] event_values) {
-        gravity[0] = alpha * gravity[0] + (1 - alpha) * event_values[0];
-        gravity[1] = alpha * gravity[1] + (1 - alpha) * event_values[1];
-        gravity[2] = alpha * gravity[2] + (1 - alpha) * event_values[2];
-
-        linear_acceleration[0] = event_values[0] - gravity[0];
-        linear_acceleration[1] = event_values[1] - gravity[1];
-        linear_acceleration[2] = event_values[2] - gravity[2];
-
-        computed_gravity = FloatMath.sqrt(gravity[0] * gravity[0] +
-                gravity[1] * gravity[1] +
-                gravity[2] * gravity[2]);
-
-        if (iterations_for_convergence > 0) {
-            // We don't use an epsilon to detect convergence because empirically
-            // devices don't seem to comply to the 0.05m/s^2 error that Google
-            // specifies.  The Motorola XT1032 seems to get ~10.10 m/s^s for gravity
-            // when it's tilted to stand on it's edge.  When laid flat,
-            // it gets pretty close with 9.89m/s^2 which is still far from 9.81m/s^2
-            iterations_for_convergence--;
-            return;
+    // Each time the baseline average changes by greater than the threshold,
+    // indicate this (see isOverThreshold) and update the baseline average to the current rolling average.
+    // So, if the threshold is 1, and the current baseline average is 10,
+    // then we need the rolling average to to reach 9 (or 11) for isOverThreshold to return true and
+    // then the baseline average is set the rolling average.
+    // This 'baseline average' method ensures that changes of acceleration can be detected over any time period.
+    class RollingAverage {
+        float[] mSamples;
+        final float mThreshold;
+        int mIdx;
+        float mTotal;
+        float mBaselineAverage;
+        boolean mIsInitialized;
+
+        public RollingAverage(int sampleCount, float changeThreshold) {
+            mSamples = new float[sampleCount];
+            mThreshold = changeThreshold;
         }
 
-        float x = linear_acceleration[0];
-        float y = linear_acceleration[1];
-        float z = linear_acceleration[2];
+        public boolean isOverThreshold(float val) {
+            mTotal -= mSamples[mIdx];
```

Comment:
Shouldn't there be some control to make sure each sample is taken over a regular time window?  It seems like this could be skewed if the accelerometers don't regularly send out metrics - which might actually be the case now that I'm typing this.